src/apps/vpn/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -258,8 +291,126 @@ bool DBusService::firewallPid(int rootpid, const QString& state) {
+  int numcaps = sizeof(newcaps) / sizeof(cap_value_t);
+  if (cap_set_flag(caps, CAP_EFFECTIVE, numcaps, newcaps, CAP_SET) ||
+      cap_set_flag(caps, CAP_PERMITTED, numcaps, newcaps, CAP_SET)) {
+    logger.warning() << "Failed to set process capability flags";
+    return;
+  }
+  if (cap_set_proc(caps) != 0) {
+    logger.warning() << "Failed to update process capabilities";
+    return;
+  }
+}
+
+/* Checks to see if the caller has sufficient authorization */
+bool DBusService::checkCallerAuthz() {
+  if (!calledFromDBus()) {
+    // If this is not a D-Bus call, it came from the daemon itself.
+    return true;
+  }
+  QDBusConnectionInterface* iface = QDBusConnection::systemBus().interface();
+
+  // If the VPN is active, and we know the UID that turned it on, as a special
+  // case we permit that user full access to the D-Bus API in order to manage
+  // the connection.
+  if (m_sessionUid != 0) {
+    QDBusReply<uint> reply = iface->serviceUid(message().service());
+    if (reply.isValid() && m_sessionUid == reply.value()) {
+      return true;
+    }
+  }
+  // Otherwise, if this is the activate method, we permit any non-root user to
+  // activate the VPN, but we will remember their UID for later authentication.
+  else if ((message().type() == QDBusMessage::MethodCallMessage) &&
+           (message().member() == "activate")) {
+    QDBusReply<uint> reply = iface->serviceUid(message().service());
+    uint senderuid = reply.value();
+    if (reply.isValid() && senderuid != 0) {
+      m_sessionUid = senderuid;
+      return true;
+    }
+  }
+  // In all other cases, the use of this D-Bus API requires the CAP_NET_ADMIN
+  // permission, which we can check by examining the PID of the sender. Note
+  // that a zero UID (root) is used as a guard value to fall back to this case.
+
+  // Get the PID of the D-Bus message sender.
+  QDBusReply<uint> reply = iface->servicePid(message().service());
+  uint senderpid = reply.value();
+  if (!reply.isValid() || (senderpid == 0)) {
+    // Could not lookup the sender's PID. Rejected!
+    logger.warning() << "Failed to resolve sender PID";
+    return false;
+  }
+
+  // Get the capabilties of the sender process.
+  cap_t caps = cap_get_pid(senderpid);
+  if (caps == nullptr) {
+    logger.warning() << "Failed to retrieve process capabilities";
+    return false;
+  }
+  auto guard = qScopeGuard([&] { cap_free(caps); });
+
+  // Check if the calling process has CAP_NET_ADMIN.
+  cap_flag_value_t flag;
+  if (cap_get_flag(caps, CAP_NET_ADMIN, CAP_EFFECTIVE, &flag) != 0) {
+    logger.warning() << "Failed to retrieve process cap_net_admin flags";
+    return false;
+  }
+  return (flag == CAP_SET);
+}
+
+uint DBusService::getCallerUid() {
+  if (!calledFromDBus()) {
+    // If this is not a D-Bus call, it came from the daemon itself.
+    return getuid();
+  }
+
+  // Get the UID of the D-Bus message sender.
+  QDBusConnectionInterface* iface = QDBusConnection::systemBus().interface();
```

Comment:
Can these be marked as `const`? 
```suggestion
  const QDBusConnectionInterface* iface = QDBusConnection::systemBus().interface();
```